### ISSUE TITLE
[explicit-length-check] not working with optional-chaining

### ISSUE THREAD

**Issue by binary64** [sindresorhus/eslint-plugin-unicorn#1263]
```
const blah = []
const a = blah.length && true // errors correctly
const b = blah?.length && true // no error given, incorrect
```

**Comment by fisker** [comment]
I don't think we intend to ignore this case, because no one should use `?.length` on array. Ignore `?.` can reducefalse positive.

**Comment by fisker** [comment]
Just like we don't check `array['length']`.

**Comment by fregante** [comment]
> because no one should use `?.length` on array

I agree that it's better to have `[]` instead of `undefined` for "unset value", but it happens

```js
function concat(strings) {
	if (!strings?.length) {
		throw TypeError('Expected array of string, received nuthin')
	}

	return string.join()
}
```

**Comment by troywoy** [comment]
This issue is pretty old and seems like it was resolved, but we're seeing the rule not correctly apply when using an optional in our TSX which in turn causes an unwanted `0` to render. We're using react-query for data fetching so our data is `ReadonlyArray<...> | undefined`.

I removed the `?` in the screenshot to show the rule is active, but in the codesandbox you can see it not applied and the unwanted `0`.

![image](https://github.com/sindresorhus/eslint-plugin-unicorn/assets/1566003/2f3d0e5e-4dd4-45f3-ab00-5becab85361e)

https://codesandbox.io/p/devbox/h45722?file=%2Fsrc%2Findex.tsx%3A9%2C18

**Comment by fregante** [comment]
If you agree with the general rule, then the fix is fine:

`data?.length > 0 && <span/>` is valid

**Comment by troywoy** [comment]
The problem with `data?.length > 0` is that then typescript complains of possibly undefined. 
![image](https://github.com/sindresorhus/eslint-plugin-unicorn/assets/1566003/b94e4ead-6aa7-434e-8cb5-b934894861dd)

That's why I gave the second type narrowing example which is what our code currently looks like. I'm not sure I agree with what was said on the thread, more so trying to understand it all. It sounds like you want to think `data` should always be a defined array? I gave a quick look for something like a `disallow-optional-length` rule but didn't see anything. Our codebase is quite large so trying to catch all these in review is going to be difficult.

**Comment by fregante** [comment]
> you want to think `data` should always be a defined array?

Yes, `unknown[] | undefined` shouldn't exist because `[]` and `undefined` are generally treated the same way in this context. So it's best to use `[]` as a null/empty value.

I know that this is an issue with complex data structures though, e.g. even just reading from a JSON file.

> The problem with `data?.length > 0` is that then typescript complains of possibly undefined.

ugh right. The "correct" solution would be `Number(data?.length) > 0` but it's not any better than what it's fixing.

The second better solution that works everywhere is `(data ?? []).length > 0`, but still not particularly an improvement.

So yeah for me this is a "bug" (or rather, an undesirable behavior) of the rule.

**Comment by viveleroi** [comment]
At the very least this rule should be able to catch `data?.length && <Foo />` because it's going to be used in real code. Our concern is that it was a path for a `0` to get rendered and while we try to avoid that, we liked having a lint rule catch it before things get to code review. Maybe you can't support `--fix` here but it appears to me like you're debating how people ideally should be writing code, not how they're really writing code.

**Comment by fregante** [comment]
Actually this issue reported that this case was *not* reported. 

Can you open a dedicated issue with a proper repro? Link to this one as an additional reference

**Comment by viveleroi** [comment]
I might be misunderstanding. The original issue is accurate to our problem, except the pseudo-code example declares `blah` to be an array when I believe they meant it could be `Array<...> | undefined` which is very common in fetch/react-query/etc uses.

```ts
const b = blah?.length && true // no error given, incorrect
```

The presence of the `?` optional chain is what causes the lint rule to miss this instance of the mistake. We *want* the lint rule to flag this use case because it's what the rule was meant to avoid.

**Comment by troywoy** [comment]
I guess I'm not sure what the takeaway from here was. Did we determine `data?.length` was *not* reported correctly and you'd like a fresh issue created stating such?

**Comment by fregante** [comment]
Please open a new bug report and fill in the fields in the new template. This is confusing now.
